plugins/Lv2Effect/Lv2Effect.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Lv2Effect.cpp - implementation of LV2 effect
+ *
+ * Copyright (c) 2018-2019 Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2Effect.h"
+
+#include <QDebug>
+#include <lv2.h>
+
+#include "AutomatableModel.h"
+#include "Lv2FxControlDialog.h"
+#include "Lv2SubPluginFeatures.h"
+#include "embed.h"
+#include "plugin_export.h"
+
+
+
+
+Plugin::Descriptor PLUGIN_EXPORT lv2effect_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"LV2",
+	QT_TRANSLATE_NOOP("Lv2Effect",
+		"plugin for using arbitrary LV2-effects inside LMMS."),
+	"Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	new Lv2SubPluginFeatures(Plugin::Effect)
+};
+
+
+
+
+Lv2Effect::Lv2Effect(Model* parent, const Descriptor::SubPluginFeatures::Key *key) :
+	Effect(&lv2effect_plugin_descriptor, parent, key),
+	m_controls(this, key->attributes["uri"])
+{
+}
+
+
+
+
+Lv2Effect::~Lv2Effect()
+{
+}
+
+
+
+
+bool Lv2Effect::processAudioBuffer(sampleFrame *buf, const fpp_t frames)
+{
```

Comment:
@DomClark  Can you please elaborate what you mean? I.e. what kind of rework are you thinking of.

Reply:
I meant that we might want to handle the effect wet/dry feature somewhere like the effect chain code, rather than re-implementing it for each effect.

Reply:
This makes sense. Any idea how this should be handled for mono effects that are being duplicated? I.e. should the D/W knob always control both D/W? For LADSPA effects, we don't seem to handle D/W in any way.

Reply:
The LADSPA dry/wet is handled when copying the audio back from the LADSPA output buffers to the LMMS buffer. The dry/wet knob currently affects both channels.
https://github.com/LMMS/lmms/blob/96cc5e0e5e9ae1277108622020b54ec1d63dd356/plugins/LadspaEffect/LadspaEffect.cpp#L243
All the native stereo plugins use a single D/W knob for both channels, and the FX mixer is flexible enough to allow the user to have different D/W amounts for each channel if they wish, so I don't see an advantage to doing anything different here.
See also #4635.

Reply:
OK. There's [lv2:wetDryRatio](http://lv2plug.in/ns/ext/parameters/#wetDryRatio) for control ports, so the plan would be that if a plugin has such a control port, we won't show that knob in the effect UI, but rather hide it from the user and connect the effect chain's D/W knob to it?

Probably, this could also be done for `lv2:gain` ("gate" in effects, we could match the dB values lv2 describes to LMMS' `0..1`) and `lv2:decay`?

Reply:
Unfortunately, LMMS's dry/wet knob goes from -1 to 1, not 0 to 1, so this isn't currently compatible with `lv2:wetDryRatio`. See #3261.

The gate and decay knobs for effects are something else entirely: if the setting "Keep effects running even without input" is disabled, then after *decay* milliseconds of  a signal below *gate*, the effect is turned off and won't be processed again until it receives new audio input. It's a mechanism designed to save processing power, although some people like to use it as an effect. I don't think it should be connected to `lv2:gain` and `lv2:decay` ports, but now that you mention it, it's implemented separately for each effect like the dry/wet knob, so should probably be implemented here. See other plugins for an example (it's the `checkGate` call that you want to look at).

Reply:
Good point with the `checkGate`.

For `lv2:dryWetRatio`, can't we simply convert that value linearly? I.e. if the user selects -1, send 0 to lv2, if they select 0, send 0.5, if they select 1, send 1?

Reply:
No. LMMS dry/wet range 0..1 should correspond exactly to LV2 drywetratio. For the negative part in LMMS, if this is kept, I think you have the options to either set it to zero or perhaps drop the sign (fabs). All three options are destructive to some degree.

Reply:
FWIW, LMMS's -1 corresponds to 2*dry-wet (extending the linear interpolation between 0 (dry) and 1 (wet)).

Reply:
That issue (#3261) and especially the linked PR there imply that it's wrong to use values `<0` for the ratio, and that the issue will be fixed anyways. So why can't the Lv2 plugin use `Effect::wetLevel()` (like `plugins/LadspaEffect/LadspaEffect.cpp`) and then convert if necessary? i.e.:

```
float w = Effect::wetLevel()
if(w<0) // ignore pointless d/w ratio < 0
{
    w = 0;
}
// now change the Lv2 d/w ratio models to use w
```

Reply:
That looks like a good way to solve this for LV2.

Reply:
Note: The `lv2:dryWetRatio` will be done in a following PR for the Parameters extension. `checkGate` will be done in a rework of this PR.